TicTacToeApp/Board.swift

```diff
@@ -0,0 +1,48 @@
+
+public class Board {
+  
+  private let markedSpaces: [Int: PlayerMark]
+  private let size = 3
+
+  public init() {
+    markedSpaces = [:]
+  }
+
+  public init(markedSpaces: [Int : PlayerMark]) {
+    self.markedSpaces = markedSpaces
+  }
+
+  public func markSpace(id: Int, mark: PlayerMark) -> Board {
+    var newSpaces = markedSpaces
+    newSpaces[id] = mark
+
+    return Board(markedSpaces: newSpaces)
+  }
+
+  public func readSpace(id: Int) -> PlayerMark? {
+    return markedSpaces[id]
+  }
+
+  public func allSpacesMarked() -> Bool {
+    return (0..<(size * size)).reduce(true, combine: { $0 && markedSpaces[$1] != nil })
+  }
+
+  public func hasWinningLine() -> Bool {
```

Comment:
The name `hasWinningLine` definitely implies knowledge of game rules. It's abstracting away the fact that three in a row makes a win.

Renaming the method would make it better, however it does not solve the root issue. The fact that rows, cols and diagonals could be used for winning combinations is part of the rules of the game. I could theoretically change the rules such that diagonals no longer win, or that the four corners result in a win.

To further illustrate this, one purely hypothetical design would be to have a `Line` class*, which responds to `areAllSpacesTheSame` or something like that. The difference being that asking if a line has three of the same mark is a data question, whereas knowledge of the rows, cols and diagonals being winning combinations is a rules one.

The question to ask is "what code would have to change if I changed <some feature>?" In this case, "some feature" would refer to various ways to win the game.

*To clarify, I'm not advocating for a `Line` class, just using it as an example.